src/test/java/com/iexec/worker/feign/LoginServiceTests.java

```diff
@@ -138,23 +141,60 @@ void shouldLogin() {
         );
     }
 
+    /**
+     * <ul>
+     *     <li>Spawn N threads
+     *     <li>Each thread tries to log in
+     *     <li>When the first thread enters the login method, it locks until N-1 threads have tried to log in
+     *     <li>In the end, a single thread should have retrieved a token
+     *     while the others should have had a "login already ongoing" message
+     * </ul>
+     */
     @Test
-    void shouldLoginOnceOnSimultaneousCalls(CapturedOutput output) throws InterruptedException, ExecutionException, TimeoutException {
-        Credentials credentials = generateCredentials();
+    void shouldLoginOnceOnSimultaneousCalls(CapturedOutput output)
+            throws InterruptedException,
+            ExecutionException,
+            TimeoutException {
+        // Pre-conditions
+        final int threadsCount = 3;
+        final ExecutorService executor = Executors.newFixedThreadPool(threadsCount);
+        final CountDownLatch gate = new CountDownLatch(threadsCount - 1);
+        final Answer<Object> waitForOtherThreads = invocation -> {
+            Awaitility.waitAtMost(1, TimeUnit.SECONDS)
+                    .pollDelay(5, TimeUnit.MILLISECONDS)
+                    .until(() -> gate.getCount() == 0);
+            return "token";
+        };
+        final Runnable loginAndDecrement = () -> {
+            loginService.login();
+            gate.countDown();
+        };
+
+        final Credentials credentials = generateCredentials();
         when(credentialsService.getCredentials()).thenReturn(credentials);
         when(coreClient.getChallenge(credentials.getAddress())).thenReturn("challenge");
-        Signature signature = SignatureUtils.hashAndSign("challenge", credentials.getAddress(), credentials.getEcKeyPair());
-        when(coreClient.login(credentials.getAddress(), signature)).thenReturn("token");
-        CompletableFuture<Void> run1 = CompletableFuture.runAsync(() -> loginService.login());
-        CompletableFuture<Void> run2 = CompletableFuture.runAsync(() -> loginService.login());
-        CompletableFuture<Void> run3 = CompletableFuture.runAsync(() -> loginService.login());
-        CompletableFuture.allOf(run1, run2, run3).get(1L, TimeUnit.SECONDS);
+        final Signature signature = SignatureUtils.hashAndSign("challenge", credentials.getAddress(), credentials.getEcKeyPair());
+        when(coreClient.login(credentials.getAddress(), signature))
+                .then(waitForOtherThreads)
+                .thenReturn("token");
+
+        // Execution
+        final CompletableFuture<?>[] completableFutures = IntStream.range(0, threadsCount)
+                .mapToObj(i -> CompletableFuture.runAsync(loginAndDecrement, executor))
+                .collect(Collectors.toList())
```

Comment:
This seems unnecessary to me. A stream has a `toArray` method.

Reply:
Thanks for the idea. I knew about this method, but couldn't make use of it last week. I finally found how to use it today. Here is the commit: https://github.com/iExecBlockchainComputing/iexec-worker/pull/587/commits/bda83c1982b67c819938c1f8baad067bfb9963b4.